### ISSUE TITLE
some cookies are obnoxious

### ISSUE THREAD

**Issue by npdoty** [npdoty/planworld#12]
Some cookies -- and it seems like maybe especially non-user-submitted ones -- express outdated, offensive or sexist attitudes. It looks like the starting set of cookies came from `fortune.txt` as distributed with VMS, and you can see [that text file in this repo](https://github.com/npdoty/planworld/blob/0e82c09515c94c28ac5d51f907e45edca339777e/support/fortune/fortune.txt).

We could:
* not show cookies on login at all
* only show user-submitted and admin-approved cookies
* take PRs on removing obnoxious fortunes and make updates to the cookies db
* something else?

**Comment by lebaer** [comment]
Thoughts:

* not showing cookies on login at all is probably the simplest way to deal with this, but I could be persuaded that this option is overkill.
* only show user-submitted and admin-approved cookies: is this cookies that are both user-submitted and admin-approved, or all user-submitted cookies plus admin-approved cookies from the fortune.txt distrib? (like, we'd remove the existing cookie DB, replace with the user-submitted cookies, and maybe add cookies back from fortune.txt as time and inclination permit?)
* taking requests on removing obnoxious fortunes: if we go this direction, we should come up with a consistent policy for guidance on removal. ("Remove on request without requiring justification" would be what I would lean towards as a welcoming community standard.) Also, what would a reasonable timeline be for end-users to expect on submitting a request?

The cookie that drove this example, per pjmorse: "A guy has to get fresh once in a while so the girl doesn't lose her confidence." I've found another round dozen I'd boot just by looking at the cookies that reference girls and women. Given Planworld's active updating population is ~75% women, having sexist, rape-culture-encouraging quotes like that as the greeting on login is something I'd like PW to shed.

Also, it bugs me that there are typos in fortune.txt, and I'd be happy to fix them as I come across them, if we don't go the "just ditch cookies" route. (I'd also be happy to remove cookies on request given the same parameters.)

(Also, how do we get this applied to PW v2 now, and PW v3 for the future?)

**Comment by pjmorse** [comment]
Another _vastly more complicated_  (but more flexible) system would allow users to flag cookies, both user-submitted and not, and any cookie with more than _n_ flags (where _n_ need not be a very large number) would be kicked out of the pool.

**Comment by npdoty** [comment]
@lebaer I believe the user-submitted cookies only appear once they are also admin-approved. (There is some admin-side functionality in place for approval of user-submitted cookies.) The already approved user-submitted cookies may not be entirely free of obnoxious comments either, but hopefully the fraction is different. To that point, we may need a policy, even a simple or implicit one, regarding what admins approve or unapprove among user-submitted cookies. I think admins can approve what in their judgment is unlikely to be obnoxious or offensive, and I'm fine with unapprove upon any good faith request with no justification required.

@pjmorse I understand that suggestion, but I don't think a flag-counting system is a feasible amount of work for this topic (at least for me), and I'm not sure flag-counting works as well in a system where selections are chosen randomly from a large corpus.

**Comment by lebaer** [comment]
I like the idea of an inline flagging system - if it could do something more simple like just email admins with a removal request and the text of the specific cookie that got flagged, that would probably be simpler to implement. Activate upon clicking text after the cookie that says something like "remove this cookie" or some other way to make it clear to users that this is how they request cookie removal. Would probably want inline confirmation after clicking, a la "You are flagging this cookie to admins for manual removal, did you mean to do that?"

**Comment by npdoty** [comment]
I've received a +1 via send for the drop-all-non-user-submitted cookies approach, based on how the fraction of cookies that's obnoxious might actually be pretty high.

I'm not clear yet on whether it's easier to drop-all or easier to add a report-this-cookie link. And we might need the report-this-cookie link in any case.

Dropping all non-user-submitted should definitely be reversible (from a backup of the database, or from the fortune.txt import procedure for starting a new planworld node), so I think it's promising.

**Comment by npdoty** [comment]
I've just seen a report of an especially obnoxious cookie that I think was user-submitted, and must have been approved by some admin in the past. So I think maybe prioritizing the report-this-cookie link makes sense.

Perhaps also we could give more people the admin privileges of reviewing and approving/unapproving cookies so that the manual review process could be more widely distributed.

**Comment by lebaer** [comment]
Resurfacing this - someone just pinged me and said they got a *Hitler* cookie on login, which is particularly egregious given it's Passover this week. I just checked the starting set of cookies and there are THREE Hitler quotes in there.

What would be the quickest way to move forward on a way for users to report said cookies - and for us to get them out of the cookie database?

**Comment by npdoty** [comment]
While this is my next planworld issue to work on, I personally don't expect to be able to turn to it for at least a week or two. I recognize that short-term fixes and fixes that could enable others (through the admin UI, say) to get involved should be prioritized.

**Comment by npdoty** [comment]
I hear the suggestion that removing the display of cookies altogether might be the way to reduce harm that involves the least development complexity, which is important given how little time we've been able to put into this.

**Comment by lebaer** [comment]
Yep, we had another report of a racist cookie today. Probably best to just remove the reference to cookies entirely as a reasonable quick fix.

**Comment by mattweberphd** [comment]
K, seems like it's time to do this. Is there any point in submitting a pull request to this repository? Or do we just pull Johnnie into this and give him line numbers to delete?

**Comment by npdoty** [comment]
This repo is not entirely current with the code running on production (and I can't put the code running in production in a public environment for the moment). But a PR that makes the change in a neat way could still be used for me or Johnnie to more easily make the same updates to the working environment.

So, I welcome advice / PRs. Or I can try to do this directly myself this weekend.

**Comment by npdoty** [comment]
Oh, disapproving all the cookies in the database was another quick solution that doesn't require any change in running code. Thanks to Johnnie (@jlodom) for that! 

Admins could whitelist acceptable cookies and keep all running code the same. (And any obnoxious cookies could be reported out of band to the admins.) Or we can still remove the cookie code altogether if we want, or add a mailto: link to report obnoxious cookies. I haven't seen the admin UI (can I add myself as an admin?), but if we add more admins and admin approval seems workable and not too burdensome, then I think that's a great easy fix and we can close this issue.

**Comment by npdoty** [comment]
I approved a handful of cookies (the only ones that I could find that were by women of color) as a test. I didn't hear a lot of feedback from users, but definitely a couple suggested we get rid of cookies altogether, and maybe a couple about approving more.
